x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java

```diff
@@ -107,6 +107,19 @@ synchronized void updatePartialResponse(
         this.reducePhase = reducePhase;
     }
 
+    /**
+     * Updates the response with the result of a partial reduction.
+     * Called only when minimize_roundtrips=true
+     * @param response
+     */
+    synchronized void updatePartialResponse(SearchResponse response) {
+        failIfFrozen();
+
+        this.responseHeaders = threadContext.getResponseHeaders();
+        this.finalResponse = response;
```

Comment:
I think that this ends up not playing well with partial results coming from the local cluster. Or do we just accept that when minimizing roundtrips with async search we lose that feature?

Also, semantic wise, it may get confusing to set this partial response to the finalResponse instance member. 

Lastly, I was trying to think of the implications/risks: similar to partially aggregated results, they are kept in memory. But search response includes hits as well, hence this may take quite a bit of memory that can't be freed until the search is completed. @original-brownbear do you see a way around this?

Reply:
On the semantic side of it, I think that get async status will return isRunning false and isPartial false if you set the partial response to the final response. It seems easy enough to fix, but maybe an indication that this may be misleading and error prone.

On the memory usage, I believe this could be a serious issue especially targeting remote clusters that may never respond (which is exactly the reason for this change). While you are able to retrieve partial results via get async search API, those will be kept in memory on the coordinating node until the search is completed (as we don't store intermediate results - see  #55464) . That's a bit of a risk as 1. the response can be big (with minimize roundtrips we get from+size hits from each cluster) 2. it may take a while, minutes, until all remote clusters have come back, and it may even happen that they don't.  I am thinking this will need some additional testing and thoughts around cleaning memory up.

Reply:
More thoughts about memory usage: the partial search response is already kept in memory via SearchResponseMerger for later merge, so perhaps this is an already existing problem. We are just adding a new reference to it in MutableSearchResponse and making it accessible from get async search API, so we may be ok. Hopefully we are not introducing a new problem. What do you both think?

Reply:
one more thought: I think that we need to reduce the response because it contains from+size hits, while the user requested size . That suggests we need a test with from higher than `0`.

Reply:
A lot to respond to here. We need slack threads for this one :-)

> I think that this ends up not playing well with partial results coming from the local cluster. Or do we just accept that when minimizing roundtrips with async search we lose that feature?

We partially lose that feature with MRT=true full merges.

With MRT=true, the local cluster also gets a CCSActionListener, so it also calls `onCcsMinimizeRoundtripsReduce(SearchResponse)` when it has finished, merging its full results with whatever other clusters have responded. 

So a typical flow is (from some logging I added to one of the IT tests):

```
Creating CCSListener for cluster [cluster_a]
Creating CCSListener for cluster []
onListShards called: shards size: 9
onPartialReduce called: shards size: 5
onFinalReduce called: shards size: 9
TSA notifyCcsMinimizeRoundtripsReduce cluster []
TSA notifyCcsMinimizeRoundtripsReduce cluster [cluster_a]
TSA createFinalResponse cluster [cluster_a]
```

So we will still have the partial aggs results from the local cluster, but only as long as no other remote cluster has returned its results. Once any full search has finished, then we ignore the partial aggs from the local cluster.

Since in `MutableSearchResponse.toAsyncSearchResponse` we create a new SearchResponse from just InternalAggs (if finalResponse is not set), we could have additional logic to merge in that SearchResponse, but that is a fair bit of work since: (1) MutableSearchResponse does not have access to a SearchResponseMerger and (2) we would have to track whether the local cluster has finished. (I guess you could tell that from when onFinalReduce is called.)

So if we wanted to do that I think it's feasible but we'd probably need a second SearchResponseMerger in the MutableSearchResponse, rather than have the MutableSearchResponse have access to the on in the TransportSearchAction.
- the Mutable

Reply:
> On the semantic side of it, I think that get async status will return isRunning false and isPartial false if you set the partial response to the final response. 

Good catch. I can adjust that by testing against the frozen flag. I'll try that now.

Reply:
> More thoughts about memory usage: the partial search response is already kept in memory via SearchResponseMerger for later merge, so perhaps this is an already existing problem. We are just adding a new reference to it in MutableSearchResponse and making it accessible from get async search API, so we may be ok. Hopefully we are not introducing a new problem. What do you both think?

Yes, SearchResponseMerger already holds all the individual cluster SearchResponses in memory. This PR creates one additional merged SearchResponse from each of those and holds that one in memory in the MutableSeachResponse, overwriting the reference to the last one. I believe @original-brownbear will be added decRef logic to MutableSearchResponse in a later PR, so I haven't done that hear.

So this PR does create one additional (possibly large) SearchResponse that is held in memory. Three options here:
1) because of Armin's shard array work on SearchResponse, this may turn out to not be much more RAM usage (?)
2) if it is more RAM usage, then we could change SearchResponseMerger to merge as each SearchResponse comes in and only hold a reference to that intermediate SearchResponse. That would align with the [original implementation](https://github.com/elastic/elasticsearch/pull/102389/files#diff-9b3d13614ee958e69c6088826f4d14821128164ae6489f2f6c917896c8af5f3dR149) I did.
3) start writing asyc-search results to disk as you outlined here: https://github.com/elastic/elasticsearch/issues/55464

Reply:
> We could rename it to latestFullResponse or mergedResponse.

I think that's a good approach. Maybe review all the usages to make sure that there is nothing else we are missing, besides the get async status flags.

> Since in MutableSearchResponse.toAsyncSearchResponse we create a new SearchResponse from just InternalAggs (if finalResponse is not set), we could have additional logic to merge in that SearchResponse, but that is a fair bit of work since: (1) MutableSearchResponse does not have access to a SearchResponseMerger and (2) we would have to track whether the local cluster has finished. (I guess you could tell that from when onFinalReduce is called.)

I can see that. There may be different solutions to this. I think it would be nice to look into it and assess how much effort it takes. It is a bit weird that you get partial results from the local cluster, and those disappear once you get results from one remote cluster. ideally we would merge the two, and you could create a new merger on the fly as well with the two responses you have? happy to brainstorm around this.

Reply:
> one more thought: I think that we need to reduce the response because it contains from+size hits, while the user requested size . That suggests we need a test with from higher than 0.

I don't think I understand what you are saying or proposing. A few responses (which may not be on the mark):

1. do you mean do some additional modification in the MutableSearchResponse rather than in the SearchResponseMerger?
2. there are already tests for SearchResponseMerger that test mergers with different from/size settings: [example](https://github.com/elastic/elasticsearch/blob/main/server/src/test/java/org/elasticsearch/action/search/SearchResponseMergerTests.java#L681). Does that cover what you are suggesting?

Reply:
I got confused around the response we are exposing. We exposed what getMergedResponse returns, hence that is properly reduced. The input search responses have from+size results, but the merged response has only size results.

Reply:
>> Since in MutableSearchResponse.toAsyncSearchResponse we create a new SearchResponse from just InternalAggs (if finalResponse is not set), we could have additional logic to merge in that SearchResponse, but that is a fair bit of work since: (1) MutableSearchResponse does not have access to a SearchResponseMerger and (2) we would have to track whether the local cluster has finished. (I guess you could tell that from when onFinalReduce is called.)

> I can see that. There may be different solutions to this. I think it would be nice to look into it and assess how much effort it takes. It is a bit weird that you get partial results from the local cluster, and those disappear once you get results from one remote cluster. ideally we would merge the two, and you could create a new merger on the fly as well with the two responses you have? happy to brainstorm around this.

I made a  first attempt at this (not pushed to this PR) was to have the SearchResponseMerger implement SearchProgressListener, so that it could get the partial results of the local cluster. But that doesn't work with the TransportSearchAction design since the SearchProgressListener is not passed in directly down the stack to the local search code. But instead, the SearchProgressListener is embedded in the task. And the AsyncSearchTask is separated by quite a few layers from the where the SearchResponseMerger is created (in fact TransportSearchAction never explicitly knows whether the search is async or not).

An alternative design could be to have the AsyncSearchTask have a reference to the same SearchResponseMerger and have it do the incremental merges, rather than the CCS ActionListener. But that would mean that sync-search and async-search for MRT=true work very differently.

So I think I could benefit from a brainstorming, design session if we want to purse this option.

The simplest solutions are:

1. Create a second SearchResponseMerger that is part of the AsyncSearchTask (or MutableSearchResponse) and can do a merger of the partial local results with whatever full SearchResponses have come in so far incrementally. This I think would be reasonable straightforward, but is fairly clunky.

2. For MRT=true, stop doing partial agg reductions for local searches entirely. One could make the argument that with MRT=true, we should treat all clusters the same - we wait until the entire result set arrives from each cluster and then do an incremental merge of those full restults. Then we don't have partial results from the local cluster showing up and then temporarily disappearing. This change would also have no negative impact on Kibana users, since Kibana does not show incremental results yet anyway.

I think my vote is actually for the last option above. I think that is cleanest - both code wise as well as in terms of explaining functionality to end users.

Reply:
I've updated the `finalResponse` field in MutableSearchResponse to be called `latestFullResponse` in [this commit](https://github.com/elastic/elasticsearch/pull/103134/commits/c7cd36293fa067b2d42aa59eb6c0261b88829e1a).

And I updated MutableSearchResponse in [this last commit](https://github.com/elastic/elasticsearch/pull/103134/commits/901c2f13f043c66009c6c58fc804e6317fe7c137) to ignore partial aggs results from the local cluster when CCS MRT=true in order to test out my suggested approach above. It passes all the automated tests I tried it against and I spent about 30 minutes manually testing local-only, CCS MRT=false and CCS MRT=true with ~15 shards per index and it looks like it behaves correctly.